README.md
# ySanic
ySanic is a [Sanic](https://sanicframework.org/) subclass with a bunch of utilities to help the developer to work with it in the context of the yRest framework (which includes [yModel](https://github.com/Garito/yModel) and [yAuth](https://github.com/Garito/yAuth) by the moment and yOpenSanic in the future)

yBlueprint adds some decorators and route creators both for CRUD and tree models

# Example
In the tests folder there is an app with two models: Todos and Nodes
Todos shows how to program the typical plain structure while Nodes shows how to program the tree ones
If you respect the contract you could use the helpers (like the routes generators)

## Installation
```pip install ySanic```

## Help
Feel free to help if you think something is weird or incomplete by submiting a pull request

### What is already needed
- [] More testing
- [] Continous integration
- [] Better help & documentation

### I'm not a technical person but still want to help
You can tip the project with cryptos too:

BTC: 1GtKxwZGR65ar9V8xafxhMiniZyqXej2GC

ETH: 0x01bd478b8C07633D2f4E58AC553f72CE4E590d56

LTC: LYUzrFX6ck5uMhw5VqcD9piQHnX7oeSLdh

XMR: 49stcvbfjEkWLjb6mdG21zMJ3uRrLmN3bazGQ8cHjjsVHYYyY61N6P7emCXhpsvB2Vc8Uuz2FA1Qk6hkE8e4ADmJQQ64eyT

ADA: DdzFFzCqrhsoUF5UjGGAYUayV5uNCJZ17PJn9V8X9MTQ26m2wDVycme42gufKufPNWMazfJLg8RKHpc1iFvn6j8BTJjaozGtLPzCDx5t

NEM: NDGYO6X3NTD6CX3V7MCCYKQPBIOYGZRXEKDLCDW2


CHANGELOG.md
# ySanic's Change Log
## 0.1.0
This version has three main differences:
- Introspection
- No decorators except notaroute
- Jinja2 emails

### Introspection
This is the biggest change since it allows several things:
- Automatic route generator
- Does the hard work for OpenApi since, in this version is key to achieve Full stack features

It uses the models' decorators to determine if a model's member is a route (when it has ```consumes``` and/or ```produces```) and what kind of route is (if it has ```consumes``` is a in route if not is and out one)

That allow us to focus in the definition of the problem's domain

### No decorators except notaroute
The only point of the decorators of ySanic was to allow to customize the response of the route but since yREST is based on REST it make sense to simplify this facet

You can always eject from yRest and use the manual routes with full control

So notaroute is the way to exclude members of the model that uses ```consumes```/```produces``` but maintain the convenience of this decorators (in terms of automatic validation and so on)

### Jinja2 emails
Anyone that has deal with emails know the nightmare they are so introducing jinja2 as a template system only simplifys what its is difficult as it is

### Generics factory, dispatcher and options
Now factory has been abstracted as a pattern in a similar way as dispatcher is

You can define which member is going to be called when a particular factory is required (which allows to have more than one type of children defined. This requieres a unique index to avoid same slug for two childs and to group them by type)

A generic_options super simple options has been defined and will be load with your routes

A generic very introspection oriented JSON encoder has been added as MyEncoder for debugging purposes. You can use it if you need it

Tests are made but, you know, test are always consider incomplet but will grow up as cases show up

### Automatic permissions when used with yAuth
It uses the yAuth's decorator ```permissions``` to map one on one with the API methods with the format ```<Model>/<member>```
It is possible to define default roles for the permissions

### Allow to load the generated routes on need
If add_routes is False on the constructor it will not load the routes (but will prepare them in case of need)
Then it is possible to load them by calling ```<ySanic object>._add_routes()


tests/app/app.py
from pymongo import ASCENDING
from sanic_mongo import Mongo, GridFS

from yModel.mongo import NotFound
from ySanic import MongoySanic

from tests.app import models
from tests.app.config import Config

class ySanicServer(MongoySanic):
  pass

def create_app(config = None):
  app = ySanicServer(models = models, name = "app")
  app.config.from_object(config or Config)

  URI = app.config.get("MONGO_URI")
  Mongo.SetConfig(app, test = URI)
  Mongo(app)
  GridFS.SetConfig(app, test_fs = (URI, "fs"))
  GridFS(app)

  @app.listener("before_server_start")
  async def setup_db(app, loop):
    table = app.mongo["test"][app.config.get("MONGO_TABLE")]

    await table.create_index([("path", ASCENDING), ("slug", ASCENDING)])

    root = app.models.Community(table)
    try:
      await root.get(path = "", name = "")
    except NotFound:
      root.load({"path": "", "name": "", "description": "The community object is the root of the site"})
      await root.create()

      admin = app.models.User(table)
      admin.load({"path": "/", "name": "Garito", "email": "garito@gmail.com"})
      await admin.create()

      await root.update({"users": [admin._id]})

  app.register_middleware(app.set_table, "request")

  if app.config.get("DEBUG", False):
    app.register_middleware(app.allow_origin, "response")

  return app

if __name__ == "__main__":
  app = create_app()
  app.run(host = app.config.get("HOST", "localhost"), port = app.config.get("PORT", 8000))


tests/testApp.py
from unittest import TestCase

from random import choice

from bson import ObjectId

from pymongo import MongoClient

from slugify import slugify

from tests.app.app import create_app

class TestTodos(TestCase):
  def setUp(self):
    self.app = create_app()
    self.client = MongoClient(self.app.config["MONGO_URI"])
    self.table = self.client.tests.tests

    self.todos = [
      {"type": "Todo", "description": "Todo 1"},
      {"type": "Todo", "description": "Todo 2"},
      {"type": "Todo", "description": "Todo 3"},
      {"type": "Todo", "description": "Todo 4"},
      {"type": "Todo", "description": "Todo 5"}
    ]

    for todo in self.todos:
      result = self.table.insert_one(todo)
      todo["_id"] = result.inserted_id

  def tearDown(self):
    for todo in self.todos:
      self.table.delete_one({"_id": todo["_id"]})

    self.client.close()

  def testNotFound(self):
    _id = ObjectId()
    req, resp = self.app.test_client.get("/todos/{}".format(str(_id)))

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], str({"_id": _id}))

  def testCreateTodo(self):
    data = {"description": "Todo's description"}
    req, resp = self.app.test_client.post("/todos", json = data)

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    doc = self.table.find_one({"_id": ObjectId(resp.json["result"])})
    self.assertEqual(doc["description"], data["description"])

    self.table.delete_one({"_id": ObjectId(resp.json["result"])})

  def testGetTodos(self):
    req, resp = self.app.test_client.get("/todos")

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    retrieved_ids = [todo["_id"] for todo in resp.json["result"]]
    expected_ids = [str(todo["_id"]) for todo in self.todos]

    self.assertListEqual(sorted(retrieved_ids), sorted(expected_ids))

  def testGetTodo(self):
    todo = choice(self.todos).copy()
    todo["_id"] = str(todo["_id"])
    req, resp = self.app.test_client.get("/todos/{}".format(todo["_id"]))

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    self.assertDictEqual(resp.json["result"], todo)

  def testGetTodoNotFound(self):
    _id = ObjectId()
    req, resp = self.app.test_client.get("/todos/{}".format(_id))

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())

    self.assertEqual(resp.json["message"], str({"_id": _id}))

  def testUpdateTodo(self):
    todo = choice(self.todos)
    data = {"description": "{} edited".format(todo["description"])}
    req, resp = self.app.test_client.post("/todos/{}".format(todo["_id"]), json = data)

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    self.assertEqual(resp.json["result"], data["description"])

  def testDeleteTodo(self):
    todo = choice(self.todos)
    req, resp = self.app.test_client.delete("/todos/{}".format(todo["_id"]))

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())
    self.assertEqual(resp.json["result"], str(todo["_id"]))

    doc = self.table.find_one({"_id": todo["_id"]})

    self.assertIsNone(doc)

class TestNodes(TestCase):
  def setUp(self):
    self.app = create_app()
    self.client = MongoClient(self.app.config["MONGO_URI"])
    self.table = self.client.tests.tests

    self.nodes = [
      # {"type": "Node", "path": "/", "name": "", "slug": "", "nodes": ["parent-1"]},
      {"type": "Node", "path": "/", "name": "Parent 1", "slug": "parent-1", "nodes": ["parent-2", "paper-2"]},
      {"type": "Node", "path": "/parent-1", "name": "Parent 2", "slug": "parent-2", "nodes": ["paper"]},
      {"type": "Node", "path": "/parent-1", "name": "Paper 2", "slug": "paper-2", "nodes": []},
      {"type": "Node", "path": "/parent-1/parent-2", "name": "Paper", "slug": "paper", "nodes": []}
    ]

    for node in self.nodes:
      result = self.table.insert_one(node)
      node["_id"] = result.inserted_id

  def tearDown(self):
    for node in self.nodes:
      self.table.delete_one({"_id": node["_id"]})

    self.client.close()

  def testMethodNotFound(self):
    node = self.nodes[1]
    url = "{}/{}/idontexist".format(node["path"], node["slug"])
    req, resp = self.app.test_client.get("/nodes{}".format(url))

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], "{} not found".format(url))

  def testGetNode(self):
    node = choice(self.nodes)
    url = "/nodes{}/{}".format(node["path"], node["slug"]) if node["path"] != "/" else "/nodes/{}".format(node["slug"])
    req, resp = self.app.test_client.get(url)

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())
    self.assertEqual(resp.json["result"]["_id"], str(node["_id"]))

  def testGetWrongNode(self):
    wrong_path = "/one/that/doesnt/exist"
    req, resp = self.app.test_client.get("/nodes{}".format(wrong_path))

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], "{} not found".format(wrong_path))

  def testGetChildren(self):
    node = self.nodes[0]
    req, resp = self.app.test_client.get("/nodes/{}/get_children".format(node["slug"]))

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    self.assertEqual(resp.json["result"][0]["_id"], str(self.nodes[1]["_id"]))
    self.assertEqual(resp.json["result"][1]["_id"], str(self.nodes[2]["_id"]))

  def testGetAncestors(self):
    node = self.nodes[-1]
    req, resp = self.app.test_client.get("/nodes{}/{}/get_ancestors".format(node["path"], node["slug"]))

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    self.assertEqual(resp.json["result"][0]["_id"], str(self.nodes[0]["_id"]))
    self.assertEqual(resp.json["result"][1]["_id"], str(self.nodes[1]["_id"]))

  def testCreateChild(self):
    node = self.nodes[-1]
    data = {"type": "Node", "name": "Child node"}
    req, resp = self.app.test_client.post("/nodes{}/{}/new/nodes".format(node["path"], node["slug"]), json = data)

    self.assertEqual(resp.status, 201)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())

    result = self.table.delete_one({"_id": ObjectId(resp.json["result"]["_id"])})
    self.assertEqual(result.deleted_count, 1)

  def testCreateChildWrongParent(self):
    wrong_path = "/one/that/doesnt/exist"
    data = {"type": "Node", "name": "Child node with wrong parent"}
    req, resp = self.app.test_client.post("/nodes{}/new/nodes".format(wrong_path), json = data)

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], "{} not found".format(wrong_path))

  def testUpdateNodeName(self):
    nodes = [
      {"type": "Node", "path": "/parent-1", "name": "Updated node", "slug": "updated-node", "nodes": ["subnode-1", "subnode-2"]},
      {"type": "Node", "path": "/parent-1/updated-node", "name": "Subnode 1", "slug": "subnode-1", "nodes": []},
      {"type": "Node", "path": "/parent-1/updated-node", "name": "Subnode 2", "slug": "subnode-2", "nodes": ["sub-subnode-1", "sub-subnode-2"]},
      {"type": "Node", "path": "/parent-1/updated-node/subnode-1", "name": "Sub subnode 1", "slug": "sub-subnode-1", "nodes": ["sub-sub-subnode"]},
      {"type": "Node", "path": "/parent-1/updated-node/subnode-1/sub-subnode-1", "name": "Sub sub subnode", "slug": "sub-sub-subnode", "nodes": []}
    ]

    for node in nodes:
      result = self.table.insert_one(node)
      node["_id"] = result.inserted_id

    node_url = "/parent-1/updated-node"
    data = {"name": "Updated updated node"}
    req, resp = self.app.test_client.put("/nodes{}/update/name".format(node_url), json = data)
    data["slug"] = slugify(data["name"])
    new_url = "/parent-1/{}".format(data["slug"])

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())
    self.assertDictEqual(resp.json["result"], data)

    docs_old = list(self.table.find({"path": {"$regex": "^{}".format(node_url)}}))
    docs_new = list(self.table.find({"path": {"$regex": "^{}".format(new_url)}}))

    self.assertFalse(docs_old)
    ids = sorted([str(doc["_id"]) for doc in docs_new])
    actual_ids = sorted([str(node["_id"]) for node in nodes[1:]])

    self.assertListEqual(ids, actual_ids)

    for node in nodes:
      self.table.delete_one({"_id": node["_id"]})

  def testUpdateWrongParent(self):
    wrong_path = "/path/that/doesnt/exists"
    req, resp = self.app.test_client.put("/nodes{}/update/nane".format(wrong_path), json = {"name": "will not work"})

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], "{} not found".format(wrong_path))

  def testDeleteNode(self):
    nodes = [
      {"type": "Node", "path": "/parent-1", "name": "Deleted node", "slug": "deleted-node", "nodes": ["subnode-1", "subnode-2"]},
      {"type": "Node", "path": "/parent-1/deleted-node", "name": "Subnode 1", "slug": "subnode-1", "nodes": []},
      {"type": "Node", "path": "/parent-1/deleted-node", "name": "Subnode 2", "slug": "subnode-2", "nodes": ["sub-subnode-1", "sub-subnode-2"]},
      {"type": "Node", "path": "/parent-1/deleted-node/subnode-1", "name": "Sub subnode 1", "slug": "sub-subnode-1", "nodes": ["sub-sub-subnode"]},
      {"type": "Node", "path": "/parent-1/deleted-node/subnode-1/sub-subnode-1", "name": "Sub sub subnode", "slug": "sub-sub-subnode", "nodes": []}
    ]

    for node in nodes:
      result = self.table.insert_one(node)
      node["_id"] = result.inserted_id

    node_url = "/parent-1/deleted-node"
    req, resp = self.app.test_client.delete("/nodes{}".format(node_url))

    children = list(self.table.find({"path": {"$regex": "^{}".format(node_url)}}))

    self.assertEqual(resp.status, 200)
    self.assertIn("ok", resp.json.keys())
    self.assertTrue(resp.json["ok"])
    self.assertIn("result", resp.json.keys())
    self.assertEqual(resp.json["result"], str(nodes[0]["_id"]))

    self.assertFalse(children)

  def testDeleteWrongParent(self):
    wrong_path = "/path/that/doesnt/exists"
    req, resp = self.app.test_client.delete("/nodes{}".format(wrong_path))

    self.assertEqual(resp.status, 404)
    self.assertIn("ok", resp.json.keys())
    self.assertFalse(resp.json["ok"])
    self.assertIn("message", resp.json.keys())
    self.assertEqual(resp.json["message"], "{} not found".format(wrong_path))


ySanic/__init__.py
from inspect import getmembers, isclass, isfunction, ismethod, iscoroutinefunction, getmro
from time import perf_counter, process_time
from pathlib import PurePath
from logging import getLogger, INFO
from functools import wraps
from smtplib import SMTP
from email.mime.text import MIMEText

import bson

from sanic import Sanic, response
from sanic.blueprints import Blueprint
from sanic.log import logger
from sanic.exceptions import InvalidUsage, MethodNotSupported

from yModel import Schema, Tree, ErrorSchema
from yModel.mongo import NotFound

from json import dumps, JSONEncoder

import re
if hasattr(re, '_pattern_type'):
  from re import _pattern_type as isPattern
else:
  from typing import Pattern as isPattern

from sanic.views import CompositionView
from typing import Type, Callable

class MyEncoder(JSONEncoder):
  def default(self, obj):
    if isclass(obj) or ismethod(obj) or isfunction(obj) or isinstance(obj, (CompositionView, frozenset, isPattern, Type, Callable, set, bson.ObjectId)):
      return str(obj)

    return JSONEncoder.default(self, obj)

class ySanic(Sanic):
  log = logger

  def __init__(self, models, **kwargs):
    add_routes = kwargs.pop("add_routes", True)
    super().__init__(**kwargs)

    self.models = models
    self._models_by_type = self._models_types()
    self._inspected, self._permissions, self._generated_routes = self._inspect(add_routes = add_routes)

    # self.log.info(dumps(self._models_by_type, indent = 2))
    # self.log.info(dumps(self._inspected, indent = 2, cls = MyEncoder))
    # self.log.info(dumps(self._generated_routes, indent = 2, cls = MyEncoder))
    # self.log.info(dumps(self.router.routes_all, indent = 2, cls = MyEncoder))

  def _models_types(self, models = None):
    trees = getmembers(models or self.models, lambda m: isclass(m) and issubclass(m, Tree))
    not_trees = getmembers(models or self.models, lambda m: isclass(m) and issubclass(m, Schema) and not issubclass(m, Tree))
    root = [name for name, model in trees]
    not_root = []
    inout = [name for name, model in not_trees]
    leafs = []

    for name, model in trees:
      for child in getattr(model, "children_models", {}).values():
        if child in root and child != model.__name__:
          del root[root.index(child)]
          not_root.append(child)
        elif child in inout:
          del inout[inout.index(child)]
          leafs.append(child)

    independent = []
    for name, model in not_trees:
      if name in inout and getmembers(model, lambda m: (ismethod(m) or isfunction(m)) and m.__module__ == model.__module__):
        del inout[inout.index(name)]
        independent.append(name)

    return {"root": root[0] if len(root) > 0 else None, "trees": not_root, "leafs": leafs, "independent": independent, "inout": inout}

  def _inspect(self, models = None, add_routes = True):
    permissions = []
    data = {}
    routes = []
    for name, model in getmembers(models or self.models, lambda m: isclass(m) and not m.__module__.startswith("yModel")):
      model_methods = getmembers(model, lambda m: (isfunction(m) or ismethod(m)) and hasattr(m, "__decorators__"))
      if model_methods:
        data[name] = {}
        if name == self._models_by_type["root"]:
          data[name]["type"] = "root"
        elif name in self._models_by_type["trees"]:
          data[name]["type"] = "tree"
        elif name in self._models_by_type["independent"]:
          data[name]["type"] = "independent"
        else:
          data[name]["type"] = "leaf"

        if data[name]["type"] in ["root", "tree"]:
          data[name]["recursive"] = model.__name__ in getattr(model, "children_models", {}).values()

        for method_name, method in model_methods:
          if "consumes" in method.__decorators__:
            dir_ = "in"
            if (hasattr(model, "factories") and method_name in model.factories.values()) or method.__decorators__["consumes"]["model"] == model.__name__:
              type_, types_ = "factory", "factories"
            else:
              type_, types_ = "updater", "updaters"
          else:
            dir_ = "out"

            if method_name in ["remove", "find_and_remove"]:
              type_, types_ = "remover", "removers"
            else:
              type_, types_ = "view", "views"

          if dir_ not in data[name]:
            data[name][dir_] = {}

          if types_ not in data[name][dir_]:
            data[name][dir_][types_] = {}

          data[name][dir_][types_][method_name] = {"method": method, "decorators": method.__decorators__}

          if not method.__decorators__.get("notaroute", False):
            route = {"model": model, "type_": type_, "is_": data[name]["type"], "data": data[name][dir_][types_][method_name]}
            routes.append(route)

            if add_routes:
              self._add_route(**route)
            if data[name]["type"] == "root" and data[name]["recursive"] and not self._models_by_type["trees"]:
              route = {"model": model, "type_": type_, "is_": "tree", "data": data[name][dir_][types_][method_name]}
              routes.append(route)

              if add_routes:
                self._add_route(**route)

          if "permission" in method.__decorators__.keys():
            context, permission = method.__qualname__.split('.')
            if permission == "__call__":
              permission = "call"
            permissions.append({"context": context, "name": permission, "description": method.__doc__,
              "roles": method.__decorators__["permission"]["default"]})

    return data, permissions, routes

  def _add_routes(self, routes = None):
    if routes is None:
      routes = self._generated_routes

    for route in routes:
      self._add_route(**route)

    if "yOpenSanic" in [class_.__name__ for class_ in getmro(self.__class__)]:
      self._route_adder("", "/openapi", "GET", self.openapi)

  def _route_adder(self, prefix, url, verb, endpoint):
    if prefix:
      url = "{}{}".format(prefix, url)

    if url not in self.router.routes_all or verb not in list(self.router.routes_all[url][1]):
      self.add_route(endpoint, url, methods = [verb])

    if url not in self.router.routes_all or "OPTIONS" not in list(self.router.routes_all[url][1]):
      self.add_route(self.generic_options, url, methods = ["OPTIONS"])

  def _add_route(self, model, type_, is_, data):
    prefix = getattr(model, "url_prefix", False)

    if type_ == "factory":
      verb = "POST"
      if is_ == "independent":
        url = "/"
        endpoint = data["method"]
      else:
        url = "/new/<as_>" if is_ == "root" else "/<path:path>/new/<as_>"
        endpoint = self.factory
    elif type_ == "updater":
      verb = "PUT"
      if is_ == "independent":
        url = "/<_id>"
        endpoint = data["method"]
      else:
        url = "/" if is_ == "root" else "/<path:path>"
        endpoint = self.dispatcher
    elif type_ == "remover":
      verb = "DELETE"
      if is_ == "independent":
        url = "/<_id>"
        endpoint = data["method"]
      else:
        url = "/<path:path>"
        endpoint = self.dispatcher
    else:
      verb = "GET"
      if is_ == "independent":
        if data["method"].__name__ == "__call__":
          url = "/<_id>"
        elif data["method"].__name__ == "get_all":
          url = "/"
        else:
          url = "/<_id>/{}".format(data["method"].__name__)
        endpoint = data["method"]
      else:
        url = "/" if data["method"].__name__ == "__call__" else "/<path:path>"
        endpoint = self.dispatcher

    self._route_adder(prefix, url, verb, endpoint)

  async def factory(self, request, path = "/", as_ = None):
    """
    The user will ask for /path/to/the/parent/new/member-list
    Where member-list is the list where the parent saves the children order
    So in the test model MinimalMongoTree the only factory will be /new/children
    """
    counter, time = perf_counter(), process_time()
    if not path.startswith("/"):
      path = "/{}".format(path)

    resp = await self.resolve_path(path)
    if resp:
      if "path" not in request.json:
        request.json["path"] = path

      method = getattr(resp["model"], resp["model"].factories[as_] if hasattr(resp["model"], "factories") and as_ in resp["model"].factories else "create_child")
      result = await method(request, as_)
      code = result.code if issubclass(result.__class__, ErrorSchema) else 201
      result = result.to_plain_dict()

      result['pref_counter'] = (perf_counter() - counter) * 1000
      result['process_time'] = (process_time() - time) * 1000
      return response.json(result, code)
    else:
      error = self.models.ErrorSchema()
      error.load({"message": "{} not found".format(path), "code": 404})
      return response.json(error.to_plain_dict(), 404)

  async def dispatcher(self, request, path = "/"):
    counter, time = perf_counter(), process_time()
    if not path.startswith("/"):
      path = "/{}".format(path)

    resp = await self.resolve_path(path, 1)

    parts = path.split("/")
    if not resp and len(parts) == 2:
      root = await self.get_root()
      resp = {"model": root, "args": parts[1]}

    if resp:
      paper = resp["model"]
      default_methods = {"GET": "__call__", "PUT": "update", "DELETE": "remove"}
      member = resp["args"] if "args" in resp else default_methods[request.method]

      method = getattr(paper, member, None)
      if method is not None and not method.__decorators__.get("notaroute", False):
        result = await method(request) if iscoroutinefunction(method) else method(request)
        code = result.code if issubclass(result.__class__, ErrorSchema) else 200
        result = result.to_plain_dict()

        result['pref_counter'] = (perf_counter() - counter) * 1000
        result['process_time'] = (process_time() - time) * 1000
        return response.json(result, code)

    error = self.models.ErrorSchema()
    error.load({"message": "{} not found".format(path), "code": 404})
    return response.json(error.to_plain_dict(), 404)

  async def generic_options(self, request, *args, **kwargs):
    return response.text("", status = 204)

  async def notify(self, notification, request, data):
    if hasattr(self, notification):
      func = getattr(self, notification)
      return await func(request, data) if iscoroutinefunction(func) else func(request, data)
    else:
      self.log.info("{}: {}".format(notification, data))

  def send_mail(self, to, subject, text = None, html = None):
    if self.config.get("DEBUG_EMAILS", False):
      self.log.info(f"to: {to}")
      self.log.info(f"subject: {subject}")
      self.log.info(f"text: {text}")
      self.log.info(f"html: {html}")
    else:
      msg = MIMEText(html, 'html')
      msg["From"] = self.config["SMTP_SENDER"]
      msg["To"] = to
      msg["Subject"] = subject

      server = SMTP("{}:{}".format(self.config["SMTP_SERVER"], self.config.get("SMTP_PORT", 587)))
      server.starttls()
      server.login(self.config["SMTP_SENDER"], self.config["SMTP_SENDER_PASSWORD"])
      server.sendmail(self.config["SMTP_SENDER"], to, msg.as_string())
      server.quit()

  async def allow_origin(self, request, response):
      response.headers["Access-Control-Allow-Origin"] = "*"
      response.headers["Access-Control-Allow-Methods"] = "GET, POST, PUT, DELETE, OPTIONS"
      response.headers["Access-Control-Allow-Headers"] = "Access-Control-Allow-Origin, Access-Control-Allow-Headers, Origin, X-Requested-With, Content-Type, Authorization"

class MongoySanic(ySanic):
  def __init__(self, models, **kwargs):
    table = kwargs.pop("table", None)
    if table is not None:
      self.table = table
    super().__init__(models, **kwargs)

  async def get_root(self):
    doc = await self.table.find_one({"path": ""})
    if doc:
      model = getattr(self.models, doc["type"])(self.table)
      model.load(doc)
      errors = model.get_errors()
      if errors:
        raise InvalidUsage(errors)
      else:
        return model
    else:
      raise NotFound("root not found")

  async def get_path(self, path):
    result = await self.get_paths([path])
    return result[0] if isinstance(result, list) else result

  async def get_paths(self, paths):
    purePaths = []
    for path in paths:
      path = PurePath(path)
      purePaths.append({"path": str(path.parent), "slug": path.name})

    result = await self.table.find_one(purePaths[0]) if len(purePaths) == 1 else await self.table.find({"$or": purePaths}).to_list(None)
    return result

  async def get_paper(self, path):
    doc = await self.get_path(path)
    if doc:
      model = getattr(self.models, doc["type"])(self.table)
      model.load(doc)
      errors = model.get_errors()
      if errors:
        raise InvalidUsage(errors)
      else:
        return model
    else:
      raise NotFound("{} not found".format(path))

  async def resolve_path(self, path, max_args = 0):
    path = PurePath(path)
    args = []
    if path.name == "":
      paper = await self.get_root()
      return {"model": paper}
    else:
      while path.name != "":
        try:
          paper = await self.get_paper(path)
        except NotFound:
          if len(args) >= max_args:
            return None
          args.append(path.name)

          path = path.parent
          continue

        if paper:
          result = {"model": paper}
          if args:
            result["args"] = args[0] if max_args == 1 and len(args) > 0 else args
          return result
        else:
          if len(args) >= max_args:
            return None
          args.append(path.name)

        path = path.parent

  async def get_file(self, filename):
    cursor = self.GridFS["test_fs"].find({"filename": filename})

    stream = None
    content_type = None
    async for file in cursor:
      content_type = file.metadata["contentType"]
      stream = await file.read()

    if stream is None and content_type is None:
      raise NotFound(filename)

    return {"stream": stream, "contentType": content_type}

  async def set_file(self, filename, data, contentType):
    await self.GridFS["test_fs"].upload_from_stream(filename, data, metadata = {"contentType": contentType})

  async def set_table(self, request):
    request.app.table = request.app.mongo["test"][request.app.config.get("MONGO_TABLE")]

def notaroute(func):
  if not hasattr(func, "__decorators__"):
    func.__decorators__ = {}
  func.__decorators__["notaroute"] = True

  @wraps(func)
  async def decorated(*args, **kwargs):
    result = await func(*args, **kwargs)
    return result

  return decorated
